Interrogate: add option to include ranks in output

vacui-dev opened the pull request:
Prior discussion: https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/2149

(Sorry for the churn -- I am not used to branch management and pull requests in GitHub)

This diff provides the option to add model's confidence as attention/emphasis

![image](https://user-images.githubusercontent.com/42512391/195226516-22010a38-546e-4d21-8878-73dd501ee4c0.png)

Why this change:

* This is particularly useful when using Interrogate DeepBooru with low "Score thresholds"
   * EG: If using Interrogate DeepBooru with a threshold of 0.5 (or even lower), you can get tags for minor scene components, which you can feed back into txt2img with the appropriate rates, to get scenes that more closely match the feel of the reference image.
* It's mostly useful for keeping minor scene components in the prompt, without completely discarding them, and without having them contribute too much emphasis / influence over the final output
* This can also give much better results when feeding the interrogate results back into either img2img or txt2img, especially when trying to generate a specific character or scene for which you have a similar concept image

Testing Steps:

* Launch Webui with command line arg: --deepdanbooru
* Navigate to img2img tab, use interrogate DeepBooru, verify tags appears as before. Use "Interrogate CLIP", verify prompt appears as before
* Navigate to Settings tab, enable new option, click "apply settings"
* Navigate to img2img, Interrogate DeepBooru again, verify that weights appear and are properly formatted. Note that "Interrogate CLIP" prompt is still unchanged

In my testing, this change has no effect to "Interrogate CLIP", as it seems to generate a sentence-structured caption, and not a set of tags.

vacui-dev committed:
Interrogate: add option to include ranks in output

Since the UI also allows users to specify ranks, it can be useful to show people what ranks are being returned by interrogate

This can also give much better results when feeding the interrogate results back into either img2img or txt2img, especially when trying to generate a specific character or scene for which you have a similar concept image

Testing Steps:

Launch Webui with command line arg: --deepdanbooru
Navigate to img2img tab, use interrogate DeepBooru, verify tags appears as before. Use "Interrogate CLIP", verify prompt appears as before
Navigate to Settings tab, enable new option, click "apply settings"
Navigate to img2img, Interrogate DeepBooru again, verify that weights appear and are properly formatted. Note that "Interrogate CLIP" prompt is still unchanged
In my testing, this change has no effect to "Interrogate CLIP", as it seems to generate a sentence-structured caption, and not a set of tags.

(reproduce changes from https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/2149/commits/6ed4faac46c45ca7353f228aca9b436bbaba7bc7)

AUTOMATIC1111 commented:
All right, so first you and another person are working on very similar things. https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/2143. Please merge and make sure yours and his stuff is compatible.

Additionally, please be reasonable. 0.23546238421344734 is just not right. 0.235 should be just as good.

vacui-dev committed:
Merge remote-tracking branch 'upstream/master' into interrogate_include_ranks_in_output

vacui-dev committed:
[1/?] [wip] Reintroduce opts.interrogate_return_ranks

looks functionally correct, needs testing

Needs particular testing care around whether the colon usage (:) will break anything in whatever new use cases were introduced by https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/2143

vacui-dev committed:
[2/?] [wip] ignore OPT_INCLUDE_RANKS for training filenames

vacui-dev committed:
[3/?] [wip] fix incorrect variable reference

still needs testing

HunterVacui commented:
No worries, I'm glad you went with the larger pull request first, I don't mind updating my changes to be compatible with master.

Working on this now, I'll make another comment when I've testing everything and make sure it works.

(I've tested and verified that the Interrogate option in img2img works, but I need to test the Training file names. Might take a while, I've never done training before.

![image](https://user-images.githubusercontent.com/42512391/195439558-ae94204b-e7ae-416d-81cd-48b2eb803e8f.png)

HunterVacui commented:
It looks like the deepbooru caption isn't being used as the filename (and the option description was changed as of https://github.com/AUTOMATIC1111/stable-diffusion-webui/commit/698d303b04e293635bfb49c525409f3bcf671dce#diff-e20573334d898aca3a51188e89e2fee20a9273a1139fabe3ec8ca762969f2293)

I created some preprocessed images and verified that the caption .txt file is being generated with deepbooru tags or CLIP tags, as configured, and that the image weights are never included in the caption file, as intended

This PR should be good to go as-is

HunterVacui commented:
*facepalm* sorry about that.. I basically recreated the diff from scratch and completely forgot to include the syntax. Luckily it should be a super fast fix
